[Tabs]Adds EdgeInsets

leonmz opened the pull request:
Adds custom edge insets for MDCTabbarView.
The scrollable view will now have 52 padding at the leading edge.
![Simulator Screen Shot - iPhone 7 - 2019-06-27 at 16 57 49](https://user-images.githubusercontent.com/16440801/60308389-50b99c00-98fd-11e9-81bf-db4192fc57c4.png)

Commit:
Adds EdgeInsets

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -311,8 +314,15 @@ - (void)layoutSubviews {
-                                                   : UIStackViewDistributionFillProportionally;
+  if (canBeJustified) {
+    self.containerView.distribution = UIStackViewDistributionFillEqually;
+    self.containerView.layoutMargins =  self.edgeInsets;
```
clang-format suggested the following change:


```
-    self.containerView.layoutMargins =  self.edgeInsets;
+    self.containerView.layoutMargins = self.edgeInsets;
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented:
bazel detected changes to the following targets:

```
//components/AppBar:SwiftExamples
//components/Tabs:Bundle
//components/Tabs:BundleFiles
//components/Tabs:ColorThemer
//components/Tabs:FontThemer
//components/Tabs:ObjcExamples
//components/Tabs:SwiftExamples
//components/Tabs:TabBarView
//components/Tabs:Tabs
//components/Tabs:Theming
//components/Tabs:TypographyThemer
//components/Tabs:private
//components/Tabs:privateTabBarView
//components/Tabs:snapshot_test_lib
//components/Tabs:snapshot_tests
//components/Tabs:snapshot_tests.swift_runtime_linkopts
//components/Tabs:snapshot_tests_test_binary
//components/Tabs:snapshot_tests_test_bundle
//components/Tabs:unit_test_sources
//components/Tabs:unit_test_tabbarview_sources
//components/Tabs:unit_test_theming_sources
//components/Tabs:unit_tests
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3.swift_runtime_linkopts
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3_test_binary
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3.swift_runtime_linkopts
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3_test_binary
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Tabs:unit_tests_IPHONE_X_IN_11_0
//components/Tabs:unit_tests_IPHONE_X_IN_11_0.swift_runtime_linkopts
//components/Tabs:unit_tests_IPHONE_X_IN_11_0_test_binary
//components/Tabs:unit_tests_IPHONE_X_IN_11_0_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented:
## API diff detected the following changes

### Tabs+TabBarView

#### MDCTabBarView

*new* property: `contentEdgeInsets` in `MDCTabBarView`

<!-- identifier: apidiff -->

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.h:
```diff
@@ -27,6 +27,9 @@ __attribute__((objc_subclassing_restricted)) @interface MDCTabBarView : UIScroll
 @property(nullable, nonatomic, strong) UITabBarItem *selectedItem;
 
+/** The content edge insets of the UITabBar. */
+@property(nonatomic) UIEdgeInsets edgeInsets;
```
Please name `contentEdgeInsets` and have them flip for RTL.  It will prevent clients from having to do the flip each time.

leonmz replied:
Done.

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -311,8 +314,15 @@ - (void)layoutSubviews {
   BOOL canBeJustified = availableWidth >= requiredWidth;
-  self.containerView.distribution = canBeJustified ? UIStackViewDistributionFillEqually
-                                                   : UIStackViewDistributionFillProportionally;
+  if (canBeJustified) {
```
Please flip for RTL here based on `effectiveUserInterfaceLayoutDirection`

leonmz replied:
Done

romoore reviewed (changes requested):
Looks good overall. Please add snapshot tests (feel free to skip RTL tests for the moment).

Commit:
Merge remote-tracking branch 'upstream/develop' into EdgeInsets

romoore commented on components/Tabs/examples/MDCTabBarViewTypicalExampleViewController.m:
```diff
@@ -84,6 +84,7 @@ - (void)viewDidLoad {
                         forState:UIControlStateSelected];
   self.tabBar.selectedItem = item1;
   self.tabBar.translatesAutoresizingMaskIntoConstraints = NO;
+  self.tabBar.edgeInsets = UIEdgeInsetsMake(0, 120, 0, 20);
```
Please pare these down to something a bit smaller. Perhaps `{0, 10, 0, 10}`?

leonmz replied:
Removed.

Commit:
Adds safe area handling to edgeInsets

Commit:
add snapshot tests

Commit:
add snapshot tests

Commit:
Merge branch 'EdgeInsets' of https://github.com/leonmz/material-components-ios into EdgeInsets

leonmz commented:
Hey Rob, I am not sure if the snapshot tests were enough.. Could you please add more according to your knowledge? Thanks a lot!

Commit:
Fix dep

romoore reviewed (changes requested):
Taking a look at this. Please don't push commits without reaching out to me first. :)

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -153,6 +162,17 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
 
+- (void)setContentEdgeInsets:(UIEdgeInsets)contentEdgeInsets {
+  UIEdgeInsets adjustedInset = contentEdgeInsets;
+  if (self.mdf_effectiveUserInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionRightToLeft) {
```
I think this flipping should be done in an internal getter. Otherwise you have to respond to changes in the semantic content attribute and keep track of the original + flipped versions.

Commit:
Docs clean-up. Swapping how the directionality of the contentEdgeInsets is used internally

Commit:
A few comments

Commit:
Fixing test to show expected behavior.

Commit:
Add dependency on MDFI18N

Commit:
Fix import for MDF

romoore commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -103,6 +103,67 @@ - (void)testItemsWithOnlyTitles {
   [self generateSnapshotAndVerifyForView:self.tabBarView];
 }
 
+- (void)testItemsWithOnlyTitlesWithInsets {
```
Needs tests for safe area insets

Commit:
Adding safeAreaInsets snapshot tests

Commit:
Update snapshot test. Add TODO

romoore reviewed (changes requested):
Looks to be currently broken. The UIStackView is being repositioned but is not switching to a Scrollable layout style.  This results in Justified tabs that can scroll.

Here's a screenshot from an iPhone XR in landscape being debugged in Reveal. It shows that the UIStackView is being positioned 44 points (the safe area inset) from the leading edge but the content didn't get switched to a scrollable configuration.

<img width="951" alt="Screen Shot 2019-06-28 at 12 10 46 AM" src="https://user-images.githubusercontent.com/1753199/60316599-32719180-9939-11e9-9748-e95cb065840e.png">

romoore commented:
Got it working by adjusting the `widthConstraint` for safe area.

|Before (`self.widthAnchor`)|After (`self.safeAreaLayoutGuide.widthAnchor`)|
|---|---|
|![Simulator Screen Shot - iPhone XR - 2019-06-28 at 00 11 56](https://user-images.githubusercontent.com/1753199/60317122-a745cb00-993b-11e9-99a5-8ebb568e0455.png)|![Simulator Screen Shot - iPhone XR - 2019-06-28 at 00 27 25](https://user-images.githubusercontent.com/1753199/60317128-ac0a7f00-993b-11e9-85dc-4c218812151f.png)|

Commit:
Removing safeAreaInsets test.

SafeAreaInsets aren't the only API used for layout - the safeAreaLayoutGuide is
also used. As a result, the snapshot tests will not accurately reflect the
behavior of the view in a real application until the safeAreaLayoutGuide can be
faked correctly.

Commit:
Remove fake superview

romoore reviewed (approved):
@leonmz All set if you want to make changes. Looks good to me in general!

romoore reviewed (changes requested):
With the recent changes for scrolling to the selected item, this PR is not ready. There are bugs when first presenting the view if the safeAreaInsets are non-zero.

romoore reviewed (changes requested):
Let's remove the `contentEdgeInsets` and focus only on the built-in padding in this PR.  We can do a second PR to make fixes for the safe area.

leonmz commented:
This PR was too heavy. Split into small different PRs.
